### ISSUE TITLE
register_authentication_servers() runs too late?

### ISSUE THREAD

**Issue by tyrann0us** [cedaro/satispress#94]
[`Authentication#register_authentication_servers()`](https://github.com/cedaro/satispress/blob/develop/src/Provider/Authentication.php#L51-L83) is a callback for the `plugins_loaded` action hook [invoked with a priority of `8`](https://github.com/cedaro/satispress/blob/develop/src/Provider/Authentication.php#L47). It adds a callback for the [`determine_current_user`](https://github.com/WordPress/WordPress/blob/5.1-branch/wp-includes/user.php#L2689-L2702) filter that is executed in [`_wp_get_current_user()`](https://github.com/WordPress/WordPress/blob/5.1-branch/wp-includes/user.php#L2643-L2711):
https://github.com/cedaro/satispress/blob/2919aabca83c2f0dca85f325b470819805e2ecf2/src/Provider/Authentication.php#L47

Now if a plugin also uses the `plugins_loaded` hook (which it should) but with a lower priority (meaning it's callback is executed earlier) that can cause problems if it calls [`wp_get_current_user()`](https://github.com/WordPress/WordPress/blob/5.1-branch/wp-includes/pluggable.php#L54-L70) (which in turn calls `_wp_get_current_user()`). 
The reason is that the `determine_current_user` filter is only run _once_ per request, i.e. if the global variable `$current_user` is still empty.

This will cause [`UnauthorizedServer#authenticate`](https://github.com/cedaro/satispress/blob/develop/src/Authentication/UnauthorizedServer.php#L26-L43) to not be executed at all, which in turn causes the `auth_status` property to be `null`, which then causes SatisPress' authentication to [fail](https://github.com/cedaro/satispress/blob/develop/src/Authentication/AbstractServer.php#L91-L93) with an `Uncaught SatisPress\Exception\HttpException: Forbidden resource requested; User: 0; URI: /satispress/packages.json in /path/to/plugins/satispress/src/Exception/HttpException.php on line 71`.

Changing the `plugins_loaded` action priority from `8` to `0` would fix this.
Is that something you might want to consider? Thanks!

**Comment by bradyvercher** [comment]
Is there a reason you need to call `wp_get_current_user()` that early? Doing so likely causes issues with more plugins than just SatisPress. For instance, the [OAuth 1 Server for the REST API](https://github.com/WP-API/OAuth1/blob/30e4339b7ff5c91817c094bf588235059e8acec7/oauth-server.php#L77) doesn't wire up its filter until `init`, which used to be earliest that functions like `wp_get_current_user()` and `get_current_user_id()` could be called.

They can be called earlier than `init` now, which caused me issues in another authentication plugin when plugins started using `get_user_locale()` on `plugins_loaded:10` after WP 4.7 was released.

**Comment by tyrann0us** [comment]
>Is there a reason you need to call `wp_get_current_user()` that early?

Don't ask me, ask the folks over at https://github.com/moderntribe. 😄
Their "plugin bootstrapper" runs the [main initialization routine](https://github.com/moderntribe/tribe-common/blob/a28bceeb3ed6ce847e8f68207990d6eb02060307/src/Tribe/Main.php#L90) at a priority of `1`:
```
add_action( 'plugins_loaded', array( $this, 'plugins_loaded' ), 1 );
```

One of their preamium addons is then initiated in the `plugins_loaded` method and calls `wp_get_current_user()` (simply put). 
I will write a ticket why they need such a low priority but I'm afraid they will change that given the fact @jesseeproductions has only recently reduced the priority  [from 5 to 1](https://github.com/moderntribe/tribe-common/commit/a28bceeb3ed6ce847e8f68207990d6eb02060307#diff-e5ad0809a1399fb28c50f3775e36ba7eR90). Note: I must have the free basic plugin installed and active to ensure that the premium addon's auto updater works.

Since SatisPress' purpose is to provide a Composer repository for non-public/premium plugins whose developers still haven't understood that composer is the de facto standard today, it must be assumed that the code quality of the plug-ins "managed" by SatisPress is also poor (CodeCanyon, ThemeForest). It can therefore not be ruled out that other plugins will call the function too early, so that the problem starts all over again.

Since `wp_get_current_user()` is [pluggable](https://codex.wordpress.org/Pluggable_Functions), maybe the best approach would be to replace it with a function that _always_ runs `determine_current_user` and or write a Trac ticket?

**Comment by tyrann0us** [comment]
Related Trac ticket: https://core.trac.wordpress.org/ticket/46586.

**Comment by bradyvercher** [comment]
Hmm, that is unfortunate. A couple of approaches I see going forward would be to:

* Attach the `determine_current_user` filter immediately. This would prevent custom authentication servers from being registered.
* Change the priority to `plugins_loaded:0`. Custom authentication servers would need to be registered earlier and it'd still be possible to wind up in the same situation if someone decides to call `wp_get_current_user()` before that.
* It might be possible to wire up `determine_current_user` immediately and iterate through the servers in that callback instead. That would likely require a bit of refactoring, but it would at least allow custom authentication servers to continue to be registered before the callback runs.

I don't think running `determine_current_user` multiple times per request is feasible since it could have unintended consequences, especially for existing authentication methods that only expect to ever be run once.

**Comment by bradyvercher** [comment]
@tyrann0us I believe this should be good to go after #97.

**Comment by tyrann0us** [comment]
I can confirm it works now for the specific plugin. Thanks alot!

**Comment by bradyvercher** [comment]
@tyrann0us Awesome! Thanks for testing that out and letting me know. I'll try to push a new release with that fix in the next day or two.
